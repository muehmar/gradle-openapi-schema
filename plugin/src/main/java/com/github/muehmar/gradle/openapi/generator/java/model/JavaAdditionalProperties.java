package com.github.muehmar.gradle.openapi.generator.java.model;

import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType.javaAnyType;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaName MAP_PROPERTY_NAME = JavaName.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, javaAnyType());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, javaAnyType());
  }

  public static JavaName additionalPropertiesName() {
    return MAP_PROPERTY_NAME;
  }

  public boolean isNotAllowed() {
    return not(allowed);
  }

  public boolean isValueAnyType() {
    return type.equals(javaAnyType());
  }

  public boolean isNotValueAnyType() {
    return not(type.equals(javaAnyType()));
  }

  public TechnicalPojoMember asTechnicalPojoMember() {
    return TechnicalPojoMember.additionalProperties(type);
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(type.getQualifiedClassName().getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
