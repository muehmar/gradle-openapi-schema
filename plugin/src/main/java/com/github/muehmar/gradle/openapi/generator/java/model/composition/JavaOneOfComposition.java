package com.github.muehmar.gradle.openapi.generator.java.model.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.composition.Assertion.assertAllObjectPojos;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.composition.OneOfComposition;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaOneOfComposition {
  private final NonEmptyList<JavaObjectPojo> pojos;
  private final Optional<JavaDiscriminator> discriminator;

  JavaOneOfComposition(NonEmptyList<JavaPojo> pojos, Optional<JavaDiscriminator> discriminator) {
    this.pojos = assertAllObjectPojos(pojos);
    this.discriminator = discriminator;
  }

  public static JavaOneOfComposition wrap(
      OneOfComposition oneOfComposition,
      Optional<Discriminator> objectPojoDiscriminator,
      PojoType type,
      TypeMappings typeMappings) {
    final Optional<JavaDiscriminator> javaDiscriminator =
        oneOfComposition
            .determineDiscriminator(objectPojoDiscriminator)
            .map(JavaDiscriminator::wrap);
    return new JavaOneOfComposition(
        oneOfComposition
            .getPojos()
            .map(pojo -> JavaPojo.wrap(pojo, typeMappings))
            .map(result -> result.getTypeOrDefault(type)),
        javaDiscriminator);
  }

  public static JavaOneOfComposition fromPojos(NonEmptyList<JavaPojo> pojos) {
    return new JavaOneOfComposition(pojos, Optional.empty());
  }

  public NonEmptyList<JavaObjectPojo> getPojos() {
    return pojos;
  }

  public PList<JavaPojoMember> getMembers() {
    return pojos
        .toPList()
        .flatMap(JavaObjectPojo::getAllMembersForComposition)
        .map(JavaPojoMember::asOneOfMember)
        .distinct(Function.identity());
  }

  public PList<TechnicalPojoMember> getPojosAsTechnicalMembers() {
    return pojos.toPList().map(TechnicalPojoMember::wrapJavaObjectPojo);
  }

  public Optional<JavaDiscriminator> getDiscriminator() {
    return discriminator;
  }
}
