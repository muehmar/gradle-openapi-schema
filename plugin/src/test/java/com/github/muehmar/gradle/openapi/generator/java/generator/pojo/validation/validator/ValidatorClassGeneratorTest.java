package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.validation.validator;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.validation.validator.ValidatorClassGenerator.validationClassGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.illegalIdentifierPojo;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo1;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo2;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfCompositions;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfCompositions;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes;
import com.github.muehmar.gradle.openapi.generator.model.composition.UntypedDiscriminator;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.constraints.PropertyCount;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class ValidatorClassGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("sampleObjectPojo2")
  void generate_when_sampleObjectPojo2_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final Writer writer =
        generator.generate(sampleObjectPojo2(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("allOfPojoWithMembers")
  void generate_when_allOfPojo_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        JavaPojos.allOfPojo(sampleObjectPojo1(), sampleObjectPojo2())
            .withMembers(
                JavaPojoMembers.fromMembers(PList.of(TestJavaPojoMembers.requiredColorEnum())));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("oneOfPojoWithMembers")
  void generate_when_oneOfPojoWithMembers_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        JavaPojos.oneOfPojo(sampleObjectPojo1(), sampleObjectPojo2())
            .withMembers(
                JavaPojoMembers.fromMembers(PList.of(TestJavaPojoMembers.requiredColorEnum())));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("oneOfPojoWithDiscriminator")
  void generate_when_oneOfPojoWithDiscriminator_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfCompositions.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()),
            UntypedDiscriminator.fromPropertyName(Name.ofString("type")));

    final JavaObjectPojo pojo = JavaPojos.oneOfPojo(javaOneOfComposition);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("oneOfPojoWithDiscriminatorAndNonStrictOneOfValidation")
  void generate_when_oneOfPojoWithDiscriminatorAndNonStrictOneOfValidation_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfCompositions.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()),
            UntypedDiscriminator.fromPropertyName(Name.ofString("type")));

    final JavaObjectPojo pojo = JavaPojos.oneOfPojo(javaOneOfComposition);

    final Writer writer =
        generator.generate(
            pojo, defaultTestSettings().withNonStrictOneOfValidation(true), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("anyOfPojoWithMembers")
  void generate_when_anyOfPojoWithMembers_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        JavaPojos.anyOfPojo(sampleObjectPojo1(), sampleObjectPojo2())
            .withMembers(
                JavaPojoMembers.fromMembers(PList.of(TestJavaPojoMembers.requiredColorEnum())));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("anyOfPojoWithDiscriminator")
  void generate_when_anyOfPojoWithDiscriminator_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaAnyOfComposition javaAnyOfComposition =
        JavaAnyOfCompositions.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()),
            UntypedDiscriminator.fromPropertyName(Name.ofString("type")));

    final JavaObjectPojo pojo = JavaPojos.anyOfPojo(javaAnyOfComposition);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("emptyPojo")
  void generate_when_emptyPojo_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo = JavaPojos.objectPojo();

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("objectPojoWithMinPropertyCountConstraint")
  void generate_when_objectPojoWithMinPropertyCountConstraint_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        JavaPojos.allNecessityAndNullabilityVariants(
            Constraints.ofPropertiesCount(PropertyCount.ofMinProperties(5)));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("objectPojoWithMaxPropertyCountConstraint")
  void generate_when_objectPojoWithMaxPropertyCountConstraint_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        JavaPojos.allNecessityAndNullabilityVariants(
            Constraints.ofPropertiesCount(PropertyCount.ofMaxProperties(8)));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("objectPojoWithNotAnyValueTypeForAdditionalProperties")
  void generate_when_objectPojoWithNotAnyValueTypeForAdditionalProperties_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        sampleObjectPojo1()
            .withAdditionalProperties(JavaAdditionalProperties.allowedFor(JavaTypes.stringType()));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("objectPojoWithNotAllowedAdditionalProperties")
  void generate_when_objectPojoWithNotAllowedAdditionalProperties_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final JavaObjectPojo pojo =
        sampleObjectPojo1().withAdditionalProperties(JavaAdditionalProperties.notAllowed());

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("illegalIdentifierPojo")
  void generate_when_illegalIdentifierPojo_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = validationClassGenerator();

    final Writer writer =
        generator.generate(illegalIdentifierPojo(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }
}
