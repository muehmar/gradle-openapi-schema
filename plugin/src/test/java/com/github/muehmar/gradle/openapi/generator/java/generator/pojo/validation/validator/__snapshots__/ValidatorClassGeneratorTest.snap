allOfPojoWithMembers=[
private class Validator {
  private boolean isColorValid() {
    return color != null;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isColorValid()
        && asSampleObjectPojo1Dto().isValid()
        && asSampleObjectPojo2Dto().isValid()
        && isAdditionalPropertiesValid();
  }
}
]


anyOfPojoWithDiscriminator=[
private class Validator {

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return getAnyOfValidCount() >= 1
        && isValidAgainstTheCorrectAnyOfSchema()
        && isAdditionalPropertiesValid();
  }
}
]


anyOfPojoWithMembers=[
private class Validator {
  private boolean isColorValid() {
    return color != null;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isColorValid()
        && getAnyOfValidCount() >= 1
        && isAdditionalPropertiesValid();
  }
}
]


emptyPojo=[
private class Validator {

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isAdditionalPropertiesValid();
  }
}
]


illegalIdentifierPojo=[
private class Validator {
  private boolean isSwitchValid() {
    return true;
  }

  private boolean isPoint_Valid() {
    return true;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isSwitchValid()
        && isPoint_Valid()
        && isAdditionalPropertiesValid();
  }
}
]


objectPojoWithMaxPropertyCountConstraint=[
private class Validator {
  private boolean isRequiredStringValValid() {
    if(requiredStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", requiredStringVal);
    }

    return false;
  }

  private boolean isRequiredNullableStringValValid() {
    if(requiredNullableStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", requiredNullableStringVal);
    }

    return isRequiredNullableStringValPresent;
  }

  private boolean isOptionalStringValValid() {
    if(optionalStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", optionalStringVal);
    }

    return isOptionalStringValNotNull;
  }

  private boolean isOptionalNullableStringValValid() {
    if(optionalNullableStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", optionalNullableStringVal);
    }

    return true;
  }

  private boolean isRequiredListWithNullableItemsValid() {
    if(requiredListWithNullableItems != null) {
      return 5 <= requiredListWithNullableItems.size()
          && requiredListWithNullableItems.size() <= 10
          && requiredListWithNullableItems.stream().allMatch(this::isRequiredListWithNullableItemsValueValid);
    }

    return false;
  }

  private boolean isRequiredListWithNullableItemsValueValid(String requiredListWithNullableItemsValue) {
    if(requiredListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", requiredListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isRequiredNullableListWithNullableItemsValid() {
    if(requiredNullableListWithNullableItems != null) {
      return 5 <= requiredNullableListWithNullableItems.size()
          && requiredNullableListWithNullableItems.size() <= 10
          && requiredNullableListWithNullableItems.stream().allMatch(this::isRequiredNullableListWithNullableItemsValueValid);
    }

    return isRequiredNullableListWithNullableItemsPresent;
  }

  private boolean isRequiredNullableListWithNullableItemsValueValid(String requiredNullableListWithNullableItemsValue) {
    if(requiredNullableListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", requiredNullableListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isOptionalListWithNullableItemsValid() {
    if(optionalListWithNullableItems != null) {
      return 5 <= optionalListWithNullableItems.size()
          && optionalListWithNullableItems.size() <= 10
          && optionalListWithNullableItems.stream().allMatch(this::isOptionalListWithNullableItemsValueValid);
    }

    return isOptionalListWithNullableItemsNotNull;
  }

  private boolean isOptionalListWithNullableItemsValueValid(String optionalListWithNullableItemsValue) {
    if(optionalListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", optionalListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isOptionalNullableListWithNullableItemsValid() {
    if(optionalNullableListWithNullableItems != null) {
      return 5 <= optionalNullableListWithNullableItems.size()
          && optionalNullableListWithNullableItems.size() <= 10
          && optionalNullableListWithNullableItems.stream().allMatch(this::isOptionalNullableListWithNullableItemsValueValid);
    }

    return true;
  }

  private boolean isOptionalNullableListWithNullableItemsValueValid(String optionalNullableListWithNullableItemsValue) {
    if(optionalNullableListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", optionalNullableListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isRequiredStringValValid()
        && isRequiredNullableStringValValid()
        && isOptionalStringValValid()
        && isOptionalNullableStringValValid()
        && isRequiredListWithNullableItemsValid()
        && isRequiredNullableListWithNullableItemsValid()
        && isOptionalListWithNullableItemsValid()
        && isOptionalNullableListWithNullableItemsValid()
        && getPropertyCount() <= 8
        && isAdditionalPropertiesValid();
  }
}
]


objectPojoWithMinPropertyCountConstraint=[
private class Validator {
  private boolean isRequiredStringValValid() {
    if(requiredStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", requiredStringVal);
    }

    return false;
  }

  private boolean isRequiredNullableStringValValid() {
    if(requiredNullableStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", requiredNullableStringVal);
    }

    return isRequiredNullableStringValPresent;
  }

  private boolean isOptionalStringValValid() {
    if(optionalStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", optionalStringVal);
    }

    return isOptionalStringValNotNull;
  }

  private boolean isOptionalNullableStringValValid() {
    if(optionalNullableStringVal != null) {
      return java.util.regex.Pattern.matches("Hello", optionalNullableStringVal);
    }

    return true;
  }

  private boolean isRequiredListWithNullableItemsValid() {
    if(requiredListWithNullableItems != null) {
      return 5 <= requiredListWithNullableItems.size()
          && requiredListWithNullableItems.size() <= 10
          && requiredListWithNullableItems.stream().allMatch(this::isRequiredListWithNullableItemsValueValid);
    }

    return false;
  }

  private boolean isRequiredListWithNullableItemsValueValid(String requiredListWithNullableItemsValue) {
    if(requiredListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", requiredListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isRequiredNullableListWithNullableItemsValid() {
    if(requiredNullableListWithNullableItems != null) {
      return 5 <= requiredNullableListWithNullableItems.size()
          && requiredNullableListWithNullableItems.size() <= 10
          && requiredNullableListWithNullableItems.stream().allMatch(this::isRequiredNullableListWithNullableItemsValueValid);
    }

    return isRequiredNullableListWithNullableItemsPresent;
  }

  private boolean isRequiredNullableListWithNullableItemsValueValid(String requiredNullableListWithNullableItemsValue) {
    if(requiredNullableListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", requiredNullableListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isOptionalListWithNullableItemsValid() {
    if(optionalListWithNullableItems != null) {
      return 5 <= optionalListWithNullableItems.size()
          && optionalListWithNullableItems.size() <= 10
          && optionalListWithNullableItems.stream().allMatch(this::isOptionalListWithNullableItemsValueValid);
    }

    return isOptionalListWithNullableItemsNotNull;
  }

  private boolean isOptionalListWithNullableItemsValueValid(String optionalListWithNullableItemsValue) {
    if(optionalListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", optionalListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isOptionalNullableListWithNullableItemsValid() {
    if(optionalNullableListWithNullableItems != null) {
      return 5 <= optionalNullableListWithNullableItems.size()
          && optionalNullableListWithNullableItems.size() <= 10
          && optionalNullableListWithNullableItems.stream().allMatch(this::isOptionalNullableListWithNullableItemsValueValid);
    }

    return true;
  }

  private boolean isOptionalNullableListWithNullableItemsValueValid(String optionalNullableListWithNullableItemsValue) {
    if(optionalNullableListWithNullableItemsValue != null) {
      return java.util.regex.Pattern.matches("Hello", optionalNullableListWithNullableItemsValue);
    }

    return true;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isRequiredStringValValid()
        && isRequiredNullableStringValValid()
        && isOptionalStringValValid()
        && isOptionalNullableStringValValid()
        && isRequiredListWithNullableItemsValid()
        && isRequiredNullableListWithNullableItemsValid()
        && isOptionalListWithNullableItemsValid()
        && isOptionalNullableListWithNullableItemsValid()
        && 5 <= getPropertyCount()
        && isAdditionalPropertiesValid();
  }
}
]


objectPojoWithNotAllowedAdditionalProperties=[
private class Validator {
  private boolean isStringValValid() {
    if(stringVal != null) {
      return java.util.regex.Pattern.matches("Hello", stringVal);
    }

    return false;
  }

  private boolean isIntValValid() {
    if(intVal != null) {
      return 10L <= intVal
          && intVal <= 50L;
    }

    return false;
  }

  private boolean isDoubleValValid() {
    if(doubleVal != null) {
      return 0 <= BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("12.5"))
          && BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("50.1")) < 0;
    }

    return false;
  }

  private boolean isValid() {
    return isStringValValid()
        && isIntValValid()
        && isDoubleValValid()
        && additionalProperties.isEmpty();
  }
}
]


objectPojoWithNotAnyValueTypeForAdditionalProperties=[
private class Validator {
  private boolean isStringValValid() {
    if(stringVal != null) {
      return java.util.regex.Pattern.matches("Hello", stringVal);
    }

    return false;
  }

  private boolean isIntValValid() {
    if(intVal != null) {
      return 10L <= intVal
          && intVal <= 50L;
    }

    return false;
  }

  private boolean isDoubleValValid() {
    if(doubleVal != null) {
      return 0 <= BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("12.5"))
          && BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("50.1")) < 0;
    }

    return false;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(String additionalPropertiesValue) {
    return additionalPropertiesValue != null;
  }

  private boolean isValid() {
    return isStringValValid()
        && isIntValValid()
        && isDoubleValValid()
        && isAllAdditionalPropertiesHaveCorrectType()
        && isAdditionalPropertiesValid();
  }
}
]


oneOfPojoWithDiscriminator=[
private class Validator {

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return getOneOfValidCount() == 1
        && isValidAgainstTheCorrectOneOfSchema()
        && isAdditionalPropertiesValid();
  }
}
]


oneOfPojoWithMembers=[
private class Validator {
  private boolean isColorValid() {
    return color != null;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isColorValid()
        && getOneOfValidCount() == 1
        && isAdditionalPropertiesValid();
  }
}
]


sampleObjectPojo2=[
com.github.muehmar.openapi.util.EmailValidator

private class Validator {
  private boolean isStringValValid() {
    if(stringVal != null) {
      return java.util.regex.Pattern.matches("Hello", stringVal);
    }

    return false;
  }

  private boolean isBirthdateValid() {
    return birthdate != null;
  }

  private boolean isEmailValid() {
    if(email != null) {
      return EmailValidator.isValid(email);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValid() {
    if(getAdditionalProperties_() != null) {
      return getAdditionalProperties_().values().stream().allMatch(this::isAdditionalPropertiesValueValid);
    }

    return false;
  }

  private boolean isAdditionalPropertiesValueValid(Object additionalPropertiesValue) {
    return true;
  }

  private boolean isValid() {
    return isStringValValid()
        && isBirthdateValid()
        && isEmailValid()
        && isAdditionalPropertiesValid();
  }
}
]